powerbi-visual-builder/src_extension/builder.ts

```diff
@@ -0,0 +1,300 @@
+      .toUpperCase()
+      .slice(-4);
+  }
+  return s;
+}
+
+function getText(url: string) {
+  return fetch(url, {
+    credentials: "include"
+  }).then(resp => resp.text());
+}
+
+class PowerBIVisualGenerator implements ExportTemplateTarget {
+  constructor(
+    public template: Specification.Template.ChartTemplate,
+    public containerScriptURL: string
+  ) {}
+
+  public getProperties() {
+    return [
+      {
+        displayName: "Visual Name",
+        name: "visualName",
+        type: "string",
+        default: "MyVisual"
+      },
+      {
+        displayName: "Description",
+        name: "description",
+        type: "string",
+        default: ""
+      },
+      {
+        displayName: "Author Name",
+        name: "authorName",
+        type: "string",
+        default: "Anonymous"
+      },
+      {
+        displayName: "Author Email",
+        name: "authorEmail",
+        type: "string",
+        default: "anonymous@example.com"
+      }
+    ];
+  }
+
+  public getFileExtension() {
+    return "pbiviz";
+  }
+
+  // Return a Promise<base64>
+  public async generate(properties: { [name: string]: any }): Promise<string> {
+    const template = this.template;
+    const config = {
+      name: properties.visualName,
+      displayName: properties.visualName,
+      description: properties.description,
+      guid: properties.visualName + randomHEX32(),
+      version: "1.0.0",
+      author: {
+        name: properties.authorName,
+        email: properties.authorEmail
+      }
+    };
+
+    const visual_json = {
+      name: config.name,
+      displayName: config.displayName,
+      guid: config.guid,
+      visualClassName: "CharticulatorPowerBIVisual",
+      version: config.version,
+      description: config.description,
+      supportUrl: "",
+      gitHubUrl: ""
+    };
+
+    const dataViewMappingsConditions: { [name: string]: any } = {};
+    const objectProperties: { [name: string]: any } = {};
+
+    // TODO: for now, we assume there's only one table
+    const columns = template.tables[0].columns as PowerBIColumn[];
+
+    for (const column of columns) {
+      // Refine column names
+      column.powerBIName = column.name.replace(/[^0-9a-zA-Z\_]/g, "_");
+      dataViewMappingsConditions[column.powerBIName] = { max: 1 };
+    }
+
+    // Populate properties
+    for (const property of template.properties as PowerBIProperty[]) {
+      property.powerBIName = (property.objectID + property.displayName).replace(
+        /[^0-9a-zA-Z\_]/g,
+        "_"
+      );
+      let type: any = { text: true };
+      switch (property.type) {
+        case "number":
+          {
+            type = { numeric: true };
+          }
+          break;
+      }
+      objectProperties[property.powerBIName] = {
+        displayName: property.displayName,
+        type
+      };
+    }
+
+    const capabilities: SchemaCapabilities = {
+      dataRoles: [
+        {
+          displayName: "Granularity (Level of Detail)",
```

Comment:
Nothing to do right now, but I've noticed that this column's overall role is hard to understand.  I'm not quite sure right now how to help clarify what this column does.